janus_core/calculations/descriptors.py

```diff
@@ -0,0 +1,139 @@
+from janus_core.helpers.log import config_logger
+from janus_core.helpers.utils import FileNameMixin, none_to_dict
+
+
+class Descriptors(FileNameMixin):  # pylint: disable=too-few-public-methods
+    """
+    Prepare and calculate MLIP descriptors for structures.
+
+    Parameters
+    ----------
+    struct : MaybeSequence[Atoms]
+        Structure(s) to calculate descriptors for.
+    struct_name : Optional[str]
+        Name of structure. Default is None.
+    invariants_only : bool
+        Whether only the invariant descriptors should be returned. Default is True.
+    calc_per_element : bool
+        Whether to calculate mean descriptors for each element. Default is False.
+    write_results : bool
+        True to write out structure with results of calculations. Default is False.
+    write_kwargs : Optional[ASEWriteArgs],
+        Keyword arguments to pass to ase.io.write if saving structure with
+        results of calculations. Default is {}.
+    log_kwargs : Optional[dict[str, Any]]
+        Keyword arguments to pass to `config_logger`. Default is {}.
+    """
+
+    def __init__(
+        self,
+        struct: MaybeSequence[Atoms],
+        struct_name: Optional[str] = None,
+        invariants_only: bool = True,
+        calc_per_element: bool = False,
+        write_results: bool = False,
+        write_kwargs: Optional[ASEWriteArgs] = None,
+        log_kwargs: Optional[dict[str, Any]] = None,
+    ) -> None:
+        """
+        Initialise class.
+
+        Parameters
+        ----------
+        struct : MaybeSequence[Atoms]
+            Structure(s) to calculate descriptors for.
+        struct_name : Optional[str]
+            Name of structure. Default is None.
+        invariants_only : bool
+            Whether only the invariant descriptors should be returned. Default is True.
+        calc_per_element : bool
+            Whether to calculate mean descriptors for each element. Default is False.
+        write_results : bool
+            True to write out structure with results of calculations. Default is False.
+        write_kwargs : Optional[ASEWriteArgs],
+            Keyword arguments to pass to ase.io.write if saving structure with
+            results of calculations. Default is {}.
+        log_kwargs : Optional[dict[str, Any]]
+            Keyword arguments to pass to `config_logger`. Default is {}.
+        """
+        self.struct = struct
+        self.struct_name = struct_name
+        self.invariants_only = invariants_only
+        self.calc_per_element = calc_per_element
+        self.write_results = write_results
+
+        if isinstance(self.struct, Sequence):
+            if any(not image.calc for image in struct):
+                raise ValueError(
+                    "Please attach a calculator to all images in `struct`."
+                )
+        else:
+            if not self.struct.calc:
+                raise ValueError("Please attach a calculator to `struct`.")
```

Comment:
I don't think this works, as `if not self.struct.calc:` will raise a `AttributeError` if it's a sequence.

We probably could convert it into a sequence, but I'm not sure we gain much. We'd need an `if isinstance(struct, Atoms)` to convert it still, so all we'd save is a couple of lines in `run`. I might think about this more for #225.

Reply:
How about: 
```suggestion
        if isinstance(self.struct, Sequence) and any(not image.calc for image in struct):
            raise ValueError(
                "Please attach a calculator to all images in `struct`."
            )

        if not isinstance(self.struct, Sequence) and not self.struct.calc:
            raise ValueError("Please attach a calculator to `struct`.")
```
My main objection isn't the line count, but more the difficulty in seeing the logic of the cases due to the nesting, else, etc. I much prefer spelling things out where possible so that I can see at a glance. Though this is purely stylistic, so I'm happy for this just to be marked resolved if you think it's better off this way.

Reply:
I think having if condition... if not condition... makes it harder to follow the logic, particularly when the condition isn't trivial, so I think I'll leave it for now.

I'll keep thinking about it, and whether to always make it a sequence. which would remove the complication anyway.